src/utils/src/arg_parser.rs

```diff
@@ -0,0 +1,567 @@
+        let mut help_builder = vec![];
+
+        help_builder.push(format!("--{}", self.name));
+
+        if self.takes_value {
+            help_builder.push(format!(" <{}>", self.name));
+        }
+
+        if let Some(help) = self.help {
+            help_builder.push(format!(": {}", help));
+        }
+        help_builder.concat()
+    }
+}
+
+/// Used for setting that `--help` flag was provided by user.
+pub enum Help {
+    Provided,
+    NotProvided,
+}
+
+/// Keeps information about the command line arguments parser.
+///
+/// * `args` - A Hash Map in which the key is an argument and the value is its associated `ArgInfo`.
+/// * `extra_args` - The arguments specified after `--` (i.e. end of command options).
+/// * `header` - This is used for storing general information about the process.
+#[derive(Debug, Default, PartialEq)]
+pub struct ArgParser<'a> {
+    args: HashMap<&'a str, ArgInfo<'a>>,
+    extra_args: Vec<String>,
+    header: String,
+}
+
+impl<'a> ArgParser<'a> {
+    pub fn arg(mut self, arg_info: ArgInfo<'a>) -> Self {
+        self.args.insert(arg_info.name, arg_info);
+        self
+    }
+
+    pub fn header(mut self, header: String) -> Self {
+        self.header = header;
+        self
+    }
+
+    pub fn help(&self) -> String {
+        let mut help_builder = vec![];
+
+        help_builder.push(format!("{}\n", self.header));
+
+        for arg in self.args.values() {
+            help_builder.push(format!("{}\n", arg.format_help()));
+        }
+
+        help_builder.concat()
+    }
+
+    // This is used for getting the value for arguments that are optional
+    // (i.e. that are neither required nor have default values).
+    pub fn value(&self, arg_name: &'static str) -> Option<String> {
+        // Safe to unwrap because we are searching in the map only
+        // for valid args.
+        self.args.get(arg_name).unwrap().user_value.clone()
+    }
+
+    pub fn extra_args(self) -> Vec<String> {
+        self.extra_args
+    }
+
+    // Splits `args` in two slices: one with the actual arguments of the process and the other with
+    // the extra arguments, meaning all parameters specified after `--` (i.e. end of command options).
+    fn split_args(args: &[String]) -> (&[String], &[String]) {
+        if let Some(index) = args.iter().position(|arg| arg == ARG_SEPARATOR) {
+            return (&args[..index], &args[index + 1..]);
+        }
+
+        (&args, &[])
+    }
+
+    // Collects the command line arguments and the values provided for them.
+    pub fn parse(&mut self) -> Result<Help> {
+        let args: Vec<String> = env::args().collect();
+
+        // Skipping the first element of `args` as it is the name of the binary, not an actual argument.
+        let (args, extra_args) = ArgParser::split_args(&args[1..]);
+        self.extra_args = extra_args.to_vec();
+
+        // If user provided the `--help` argument, we want to stop the parsing.
+        if args.contains(&HELP_ARG.to_string()) {
```

Comment:
The user should decide on how to treat any args, including `--help`, and the parser should continue to do its job, not stop here.

Consider, for instance, a subcommand scenario like `command  subcommand --help`. In this case, the user might choose to show specific help information regarding `subcommand`, so they might need the whole command line parsed (i.e. to identify the subcommand).

The parser should just parse the information and validate it against the provided schema. It's not its job to interpret the meaning of the information it parses.